[ypuppy] iP

ypuppy opened the pull request:
### LUMI 



> shine Brighter, Organize Smarter! ✨ -- LUMI TEAM

_Lumi is your ultimate task manager, designed to help you organize and track your tasks efficiently._
 - User-friendly and intuitive
- Lightweight and fast
- Reliable with automatic data persistence
- It's completely FREE for students in NUS.

**Getting Started** 🚀

Just follow these simple steps:

1. Download Lumi from [here](https://github.com/ypuppy/ip/releases/tag/v0.2)
2. Run the application, you can find details here[downloadLumi.txt](https://github.com/user-attachments/files/18688523/downloadLumi.txt)
3. Start adding your tasks.
4. Let Lumi handle the rest!


_Features_

- [ ] Managing tasks 📋
- [ ] Saving and loading tasks automatically 💾
- [ ] Deadline task 
- [ ] Event task
- [ ] Reminders (coming soon)
- [ ] Task prioritization (coming soon!)


If you're a Java programmer, you can customize Lumi. Here’s the `main` method to get started:

```
public` class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

**More About Lumi**

Lumi stores tasks persistently using file-based storage. It follows a structured format, ensuring that your tasks are never lost. You can modify Lumi using Java's object-oriented programming features, such as inheritance and polymorphism.

Want to contribute? Check out the [Lumi GitHub repository](https://github.com/ypuppy/ip.git) and start coding!

Try Lumi today and make your task management effortless! ✨

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
add greeting and exit

Commit:
add echo function and changed the name of chatbot

Commit:
added a List

Commit:
added mark as done feature, new class Task.java

Commit:
add three type of Task, modified Lumi.java

Commit:
added testcase for level 0 to 4

Commit:
added LumiException

Commit:
added Exceptions for empty task prompt and empty ddl

Commit:
added delete feature, modified input flexibility

Commit:
Add FileManager.java

Commit:
Merge branch 'branch-Level-7'
Add save to file feature

Commit:
Update date handling

Commit:
Merge branch 'branch-Level-8'
Update Lumi.java to handle date

Commit:
Add Parser.java Ui.java TaskList.java

Commit:
Add more OOP

Commit:
Merge branch 'A-MoreOOP'

Commit:
Moved into package lumi

Commit:
Merge branch 'branch-A-Packages'

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add Junit testing

Commit:
Update Junit test

Commit:
Fixed

Commit:
Add find feature

Commit:
Merge branch 'branch-Level-9'

Commit:
Update UnmarkCommand.java

Commit:
Updated UnmarkCommand.java

Commit:
Add javaDoc

Commit:
Merge branch 'branch-A-javaDoc'

Commit:
Update with CodingStandard

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Add checkstyle

Commit:
Update checkstyle

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add MainWindow.java

Commit:
Add initial Fxml GUI

AbiHalim commented on src/main/java/lumi/EventCommand.java:
```diff
@@ -0,0 +1,31 @@
+package lumi;
```
Any reason why all Commands and Tasks are in the same `lumi` package instead of organizing in a separate package?

AbiHalim commented on src/main/java/lumi/Task.java:
```diff
@@ -0,0 +1,63 @@
+//to muatate the Task status while marking, have to disable final modifier
```
Should this comment be removed?

AbiHalim commented on src/main/java/lumi/Storage.java:
```diff
@@ -0,0 +1,135 @@
+package lumi;
+
+//import java.io.*;
```
Can this line be removed?

AbiHalim commented on src/main/java/lumi/Ui.java:
```diff
@@ -0,0 +1,116 @@
+     * @param tasks The task list containing previously saved tasks.
+     */
+    public void showWelcome(TaskList tasks) {
+        String logo = " __\n"
+                + "| |    _   _ __ __ __ _________\n"
+                + "| |   | | | | |\\ \\/ /| |__  __|\n"
+                + "| |__ | |_| | | \\__/ | |_|  |_|\n"
+                + "|____/ \\__,_|_|      |_|______|\n";
```
I like how you put the logo in its own variable 👍

AbiHalim commented on src/main/java/lumi/Ui.java:
```diff
@@ -0,0 +1,116 @@
+
+    }
+
+    /**
+     * Displays a horizontal line for UI formatting.
+     */
+    public void showLine() {
+        System.out.println("____________________________________________________________");
+    }
```
I like how you made showLine() into its own method

AbiHalim commented on src/main/java/lumi/FindCommand.java:
```diff
@@ -0,0 +1,43 @@
+ * Represents a command to find tasks containing a specific keyword.
+ */
+public class FindCommand extends Command {
+    private final String keyword;
```
If the variable is final, should you make the name in SCREAMING_SNAKE_CASE?

AbiHalim reviewed (commented):
LGTM. Just a few nits to fix.

Commit:
Update GUI

ho-thanh-und commented on src/test/java/lumi/CommandTest.java:
```diff
@@ -0,0 +1,98 @@
+    @Test
+    void testListCommandWithTasks() throws LumiException {
+        tasks = new TaskList();
+        tasks.getTasks().add(new lumi.Todo("Read a book"));
+        tasks.getTasks().add(new lumi.Deadline("Submit report", "2023-12-10"));
+
```
Very detailed testing!. I notice that for other segments in this file, you used the addTasks() method, but for this why not? I think using addTasks() can better ensure **encapsulation**!

ho-thanh-und commented on src/main/java/lumi/TaskList.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @return {@code true} if the task list is empty, otherwise {@code false}.
+     */
+    boolean isEmpty() {
+        return tasks.isEmpty();
+    }
+
+    /**
+     * Returns the number of tasks in the list.
+     *
+     * @return The size of the task list.
+     */
+    int size() {
+        return tasks.size();
+    }
+
+    /**
+     * Retrieves a task at the specified index.
+     *
+     * @param index The index of the task.
+     * @return The task at the specified index.
+     */
+    Task get(int index) {
+        return tasks.get(index);
+    }
```
Maybe you could add a public access modifier for better method visibility?

ho-thanh-und reviewed (commented):
I loved how you structured your codebase, made it very intuitive to understand (like dividing commands separately). Very few errors imo and had a fun time reviewing your PR!

Commit:
Update MainWindow

Commit:
Merge branch 'branch-Level-10'

Commit:
Update MainWindow

Commit:
Add assertions to Parser for improved input validation

- Added assertions to ensure and non-empty user input.
- Verified command structure.
- Ensured TaskList, UI, and Storage instances are valid before processing.
- Improved error handling to prevent unexpected crashes.

Commit:
Remove magic numbers in Parser class

Magic numbers were present in the Parser class, making the code
harder to read, maintain, and modify. Values like the minimum
word length for different commands and the expected date length
were directly used in conditions, reducing clarity.

Introduce static final constants for minimum word length
requirements (`MIN_WORD_LENGTH_FOR_GENERAL_COMMAND`,
`MIN_WORD_LENGTH_FOR_DEADLINE`, `MIN_WORD_LENGTH_FOR_EVENT`)
and minimum date length (`MIN_DATE_LENGTH`). Replace all
occurrences of hardcoded numbers with these constants.

Using constants improves code readability, maintainability,
and reduces the risk of errors. It also makes future changes
easier since adjustments can be made in a single location.

ypuppy committed:
Merge pull request #1 from ypuppy/branch-A-Assertions

Branch a assertions

The Parser class lacked assertions to verify critical assumptions, 
which could lead to unexpected behavior if invalid inputs were passed. 
Without these checks, debugging errors related to command parsing 
could be more difficult.

Introduce assertions to validate that input strings are non-null 
and non-empty before processing. Ensure that required parameters 
for commands are correctly formatted and present before execution.

Adding assertions improves robustness by catching invalid inputs 
early in development. It helps maintain code correctness while 
providing clear failure points when assumptions are violated.

This merge ensures better input validation without affecting 
existing logic or functionality.

ypuppy committed:
Merge branch 'master' into branch-A-CodeQuality

ypuppy committed:
Merge pull request #2 from ypuppy/branch-A-CodeQuality

Branch a code quality

Magic numbers were present in the Parser class, making the code 
harder to read, maintain, and modify. Values like the minimum 
word length for different commands and the expected date length 
were directly used in conditions, reducing clarity.

Introduce static final constants for minimum word length 
requirements (`MIN_WORD_LENGTH_FOR_GENERAL_COMMAND`, 
`MIN_WORD_LENGTH_FOR_DEADLINE`, `MIN_WORD_LENGTH_FOR_EVENT`) 
and minimum date length (`MIN_DATE_LENGTH`). Replace all 
occurrences of hardcoded numbers with these constants.

Using constants improves code readability, maintainability, 
and reduces the risk of errors. It also makes future changes 
easier since adjustments can be made in a single location.

Commit:
Correct formatting to align with coding standards

After resolving merge conflicts in the Parser class, some parts
of the code no longer adhered to the project's coding standards.
Inconsistent indentation, spacing, and line breaks made the code
less readable and harder to maintain.

Fix formatting issues by aligning indentation, correcting spacing,
and ensuring proper line breaks where needed. Maintain consistency
with the project's established coding style.

Applying proper formatting improves readability, enhances
maintainability, and ensures a cleaner codebase that is easier
to navigate and modify.

Commit:
Update Parser Storage and Event file

Current Storage file does not handle Event type
Event class does not handle string to date
Parser class has bugs that string is split
in a way that may cause undefined behaviour
mainly bacause we split by every space

Change split(" ") to split(\\s+), add  limit 2
Add DateParse in Event class

Limit 2 in first split to avoid extra parts
in words[] due to
user input that may have extra space

Commit:
Add tag using stream

Previously, tasks could not be categorized or labeled,
making it difficult to organize them based on
user-defined themes or priorities.

Introduce a `tag` command that allows users to
add tags to tasks.
to persist tags when saving and loading tasks.
Update `Parser` to handle the
new `tag` command and validate input.

Using `Optional<String>` for tags ensures
that tasks without tags remain clean.
The new feature improves task organization without
affecting existing functionality.

ypuppy committed:
Merge branch 'master' into branch-A-Streams

ypuppy committed:
Merge pull request #3 from ypuppy/branch-A-Streams

Add tag using stream

Previously, tasks could not be categorized or labeled, making it
difficult to organize them based on user-defined themes or priorities.
Users had no way to associate additional context with tasks.

Introduce a `tag` command that allows users to add tags to tasks
using the format `tag <task number> #<tag>`. Tags are stored in the
`Task` class and displayed alongside task details. Modify `Storage`
to persist tags when saving and loading tasks. Update `Parser` to
handle the new `tag` command and validate input.

Using `Optional<String>` for tags ensures that tasks without tags
remain clean. The new feature improves task organization without
affecting existing functionality.

Commit:
Update Tag feature

The Parser class had some bug when handling Tag function
the toString method formate is messed

Update Parser spliting and error handling
Update the to String in every Task class

Commit:
Update GUI using css

Commit:
Add personality to Lumi

Commit:
Add bot and user image

Commit:
Make the codebase clean

Commit:
Add more tests for basic command and UI.png

after changing personality and updating files
with more error handling
test file needs to be update in order to test
the new lumi

Add EventCommand testing

Commit:
Add README

Commit:
Update cli test

Commit:
Add more asserion

Commit:
Update README

Commit:
Final Update README

Commit:
Adjust file name of Ui.png

Commit:
Improve code quality
